### PR TITLE
Generic hidden state for RecurrentPPO

### PR DESCRIPTION
**Pull request by rhaps0dy**
(This PR is after #7 and #8 , which I factored out to make review easier.)

Add the remaining parts for generic-state recurrent PPO.

- The policy in `common/recurrent/policies.py` is still based on LSTMs, but now uses the `recurrent_initial_state(...)` interface to indicate what its hidden state is.
- `RecurrentPPO` is fully generic over hidden state types.
- Fixed all the type errors
  - Also re-enabled mypy for the whole codebase

### CODE REVIEW

File: stable_baselines3/ppo_recurrent/ppo_recurrent.py [hunk 1]
```diff
@@ -375,23 +371,21 @@ def train(self) -> None:
                     # Convert discrete action from float to long
                     actions = rollout_data.actions.long().flatten()
 
-                # Convert mask from float to bool
-                mask = rollout_data.mask > 1e-8
-
```
Comment (rhaps0dy) [inline stable_baselines3/ppo_recurrent/ppo_recurrent.py]: The `rollout_data.mask` is now already a bool

File: stable_baselines3/ppo_recurrent/ppo_recurrent.py [hunk 2]
```diff
@@ -260,7 +257,7 @@ def collect_rollouts(  # type: ignore[override]
 
         callback.on_rollout_start()
 
-        lstm_states = deepcopy(self._last_lstm_states)
+        lstm_states = non_null(self._last_lstm_states)
```
Comment (rhaps0dy) [inline stable_baselines3/ppo_recurrent/ppo_recurrent.py]: It's not actually necessary to copy each tensor. They don't get overwritten.

File: .circleci/config.yml [hunk 3]
```diff
@@ -9,7 +9,7 @@ parameters:
   docker_img_version:
     # Docker image version for running tests.
     type: string
-    default: "a0d53ea"
+    default: "03a594c"
```
Comment (rhaps0dy) [inline .circleci/config.yml]: This is a more recent image -- a few dependencies were added, though they don't impact this codebase (only `learned-planners`).

File: .circleci/config.yml [hunk 4]
```diff
@@ -51,7 +51,7 @@ jobs:
           command: ruff .
       - run:
           name: Typecheck (mypy)
-          command: mypy --exclude '^stable_baselines3/common/recurrent/policies\.py$' stable_baselines3/common tests
+          command: mypy .
```
Comment (rhaps0dy) [inline .circleci/config.yml]: Start typechecking all the things again!

File: stable_baselines3/common/recurrent/policies.py [hunk 5]
```diff
@@ -209,42 +221,73 @@ def _process_sequence(
...
+            lstm_states_vf = (lstm_states_pi[0].detach(), lstm_states_pi[1].detach())
+        return ((latent_pi, latent_vf), RNNStates(lstm_states_pi, lstm_states_vf))
+
+    def _recurrent_latent_vf_from_features(
+        self, vf_features: th.Tensor, state: RNNStates, episode_starts: th.Tensor
+    ) -> Tuple[th.Tensor, Optional[LSTMStates]]:
+        "Get only the vf features, not advancing the hidden state"
+        if self.lstm_critic is None:
+            if self.shared_lstm:
+                with th.no_grad():
```
Comment (dan-pandori) [inline stable_baselines3/common/recurrent/policies.py]: I would have thought that we need this `with th.no_grad():` at the top level (ie, at line 257 and applying to all parts of this function).

In particular, I'm wondering if we might accidentally alter gradients on line 266 otherwise.

File: stable_baselines3/ppo_recurrent/ppo_recurrent.py [hunk 6]
```diff
@@ -148,14 +146,38 @@ def __init__(
...
+                batch_size > 1
+            ), "`batch_size` must be greater than 1. See https://github.com/DLR-RM/stable-baselines3/issues/440"
+
+        if self.env is not None:
+            # Check that `n_steps * n_envs > 1` to avoid NaN
+            # when doing advantage normalization
+            buffer_size = self.env.num_envs * self.n_steps
+            assert buffer_size > 1 or (
+                not normalize_advantage
+            ), f"`n_steps * n_envs` must be greater than 1. Currently n_steps={self.n_steps} and n_envs={self.env.num_envs}"
```
Comment (dan-pandori) [inline stable_baselines3/ppo_recurrent/ppo_recurrent.py]:
```
f"`n_steps * n_envs` must be greater than 1 when `normalize_advantage` is true.
```
etc

**Review by dan-pandori — APPROVED** [review]
Sorry for the delayed review, I am still only firing on like half of my cylinders (so to speak).
